refactor for more profound use of Lab

paolochiodi opened the pull request:
Trying to refactor to use Lab experiments and tests instead of the long list of functions in the waterfall.
This should give us better reporting and access to all the tool Lab provides

paolochiodi commented:
Things I like in the current state: 
- proper reporting

Things I don't like:
- tests still dependant on order - need to add lot of setup code to make them order independent
- how I pass the "environment" to the tests still feels an hack

@rjrodger what do you think about this?

paolochiodi commented:
I've finally ported all the tests (with some refactoring and bug fixing) to the new style.
Also started using them (also good as example) on the mysql store: https://github.com/mirceaalexandru/seneca-mysql-store/pull/20

I'm waiting for @rjrodger approval on the new style before merging

paolochiodi commented:
Note: tests still order dependant, will address that in a follow-up pr

colmharte commented:
Not backwards compatible so would require bump to major version number

geek commented on store-test.js:
```diff
@@ -1,17 +1,14 @@
-  readline.moveCursor(process.stdout, 55, -1)
-  console.log("PASSED")
-}
+var Lab = require('lab');
```
feel free to lowercase... this is a defensive style we use in hapi land to clearly indicate our modules.

geek commented on store-test.js:
```diff
@@ -27,567 +24,492 @@ var bartemplate = {
-          scratch.foo1.p2 = 'v2'
+      foo1.save$(verify(done, function (foo1) {
+        assert.isNotNull(foo1.id)
+        assert.equal('v1',foo1.p1)
```
could use a space after comma

geek commented on store-test.js:
```diff
@@ -27,567 +24,492 @@ var bartemplate = {
-  }
-}
-
+function verify (cb, tests) {
```
Might consider removing and letting tests fail without the catch.  I know its not as DRY, but sometimes its easier to follow test code when its all contained within the same function.

paolochiodi replied:
On a general note I agree. 
However seneca seems to catch exceptions that happen in a callback to one of the entity methods. This will hide actual tests failures (exceptions raised by assertions) because failures will always be reported as generic timeouts

geek commented on store-test.js:
```diff
@@ -27,567 +24,492 @@ var bartemplate = {
+function basictest (settings) {
+  var si = settings.seneca
+  var must_merge = !!settings.must_merge
+  var script = settings.script || Lab.script()
```
May want to move to the top of the file, in order to contain all of the tests for this file in a single script variable.

paolochiodi replied:
As discussed on chat, this is the point of the new approach. 
We need to share the script (and other settings, such as seneca) among other test suites including ones outside this module (and provided by store implementations).
We will follow this approach of passing the script around - as lab doesn't have a default way to do this.
I'll consider splitting the different suites in different files.

paolochiodi commented:
I've removed the WIP "tag". The refactor to use Lab is completed - and I've fixed a couple of bugs along the way.  
Wyatt confirmed this is a good approach.

I've bumped version to 0.4.0. As this change is not backward compatible we may want to bump the major version instead, but I'm not sure @rjrodger want to get into 1.0.0.
A part from that I'm happy with the PR.

I'll work on two following PRs:
- one to split tests in different files
- one to add tests (get specific with the remove tests and check compliance with https://github.com/rjrodger/seneca/blob/master/doc/data-store.md)
